[Centralized CLI] Requirements

thomasdanan opened the pull request:
Centralized CLI requirements (Draft)

**Component**:

'doc', 'requirements'

**Context**: 

Centralized CLI requirements

Closes: #2322

thomasdanan committed:
requirements: high level principles

First draft for centralizd CLI requirements. Focus on High Level 
Principles for now.

Closes: #2322

bert-e commented:
# Hello thomasdanan,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+- Cluster Service Administration (Configure Dex, Prometheus, Alert Manager, ...)
+
+K8s provides the kubctl CLI, which is distributed and enables all kind of
+interaction with all resources available in etcd DB but its usage often requires
```
Not sure you want to talk about etcd here as kubectl basically just interact with apiserver he don't really care about etcd

sayf-eddine-scality replied:
apiserver is just a giant wrapper in front of the etcd cluster anyway

TeddyAndrieux replied:
For the moment yes but maybe one day they will use another backend :)

thomasdanan replied:
yes agree @TeddyAndrieux will change to apiserver

thomasdanan replied:
done

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+K8s provides the kubctl CLI, which is distributed and enables all kind of
+interaction with all resources available in etcd DB but its usage often requires
+to build verbose yaml files. Also kubectl tool only expose the standard K8s
+resources but not the custom ones, such as Volumes.
```
> kubectl tool only expose the standard K8s resources but not the custom ones, such as Volumes.

Hmm, he do. You can do `kubectl get volume`

slaperche-scality replied:
yeah, you can manipulate custom resources with kubectl.

thomasdanan replied:
okay my bad ... will remove this

thomasdanan replied:
done

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+
+Currently, MetalK8s provides other set of CLI or manual procedures, but those
+are located in various locations, their usage may vary and they are not
+developed using the same technologies.
```
IMO we don't want to talk about technologies but logic here.
```suggestion
developed using the same logic.
```

thomasdanan replied:
done

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+administrators, it will be used as a standard to follow for all other MetalK8s
+CLIs:
+- CLI follows kubctl style
+- CLI is not interactive
```
except for some password that should be asked at the very beginning of the script

thomasdanan replied:
done

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+CLIs:
+- CLI follows kubctl style
+- CLI is not interactive
+- CLI should not require password input
```
Except for user administration right ?

Ebaneck replied:
When provisioning new OIDC static users we mandatorily need to provide password hashes so I accept with @TeddyAndrieux that user administration should be an exception here.

TeddyAndrieux replied:
Yes, but like I said in my previous comment this should be at the very beginning of the script

thomasdanan replied:
done

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+- CLI implementation relies on secured APIs
+- CLI support completion for easy discovery
+- CLI output is standardized (yaml, json, ...)
+- CLI can be executed from outside of the cluster
```
Not all CLI

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+- CLI should not require password input
+- CLI implementation relies on secured APIs
+- CLI support completion for easy discovery
+- CLI output is standardized (yaml, json, ...)
```
IMO we should follow kubectl style so standardized yes but not yaml or json by default
Something "pretty" and human readable by default with an option to change it to yaml, json, ...

slaperche-scality replied:
`s/yaml/YAML/` and `s/json/JSON/`

thomasdanan replied:
done

TeddyAndrieux reviewed (dismissed):
Do we plan to have new CLI only in metalk8s-2.6 ?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

sayf-eddine-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+This makes the CLI and associated documentation not super intuitive and it
+also makes the maintenance more expensive in the long term.
+
+The goal of the project is to provide MetalK8s administrator an intuitive and
```
```suggestion
The goal of the project is to provide MetalK8s administrator with an intuitive and
```

thomasdanan replied:
done

sayf-eddine-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+- CLI is not interactive
+- CLI should not require password input
+- CLI implementation relies on secured APIs
+- CLI support completion for easy discovery
```
on medium to large platforms it becomes super slow if this feature supports autocompleting pods/svc/volumes names. this needs to be counted for

TeddyAndrieux replied:
Hmm for me we definitly not want completion for object name but just for different operation we can do with the CLI

slaperche-scality replied:
Hum, completion with <TAB> can be shell-specific (for the different operation)
I'm not sure we want to thread into this direction.

I think it's easier and as important to have a useful `-h/--help`, like `foo -h` list the top level commands, then a `foo command -h` gives the help of `command` and so on.
Like what `kubectl` (and others) does.

Ebaneck replied:
For me, autocompletion for both object names and operations is a nice to have. Btw, @sayf-eddine-scality what is the cause of the slowness when autocompleting pods/svc/volumes names??

sayf-eddine-scality replied:
https://github.com/kubernetes/kubernetes/issues/82883
loading a huge number of objects and then parsing their metadata takes some time (counting also network latency) and this happens every time you press Tab.

TeddyAndrieux replied:
Auto completion for objects name is a bad idea, because to discover btw you need to access apiserver so you likely need a kubeconfig ... and yes if you have tones of objects ect ...

> I think it's easier and as important to have a useful -h/--help, like foo -h list the top level commands, then a foo command -h gives the help of command and so on.
Like what kubectl (and others) does.

Agree, personnaly think having nice help text is good enough but with `cmd` lib from python we can have also an interpreter with completion

slaperche-scality replied:
> Agree, personnaly think having nice help text is good enough but with cmd lib from python we can have also an interpreter with completion

Ha I was thinking about `argparse`, but if we have a lib that does it for us then let's have completion on subcommands :slightly_smiling_face:

Ebaneck replied:
> Auto completion for objects name is a bad idea, because to discover btw you need to access apiserver so you likely need a kubeconfig ... and yes if you have tones of objects ect ...
> 
If we provide OIDC login creds at the very beginning of the script as you say then we can use the resulting OIDC auth token to generate a kubeconfig and use for subsequent operations

> > I think it's easier and as important to have a useful -h/--help, like foo -h list the top level commands, then a foo command -h gives the help of command and so on.
> > Like what kubectl (and others) does.
> 
> Agree, personnaly think having nice help text is good enough but with `cmd` lib from python we can have also an interpreter with completion

slaperche-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+- Solution Administration (CRUD Environment, Import/Remove Solution, ...)
+- Cluster Service Administration (Configure Dex, Prometheus, Alert Manager, ...)
+
+K8s provides the kubctl CLI, which is distributed and enables all kind of
```
`s/kubctl/kubectl`, BTW `kubectl` is described as a `command line tool ` not `CLI`

thomasdanan replied:
I think it is okay and easier to talk about CLI

slaperche-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+
+K8s provides the kubctl CLI, which is distributed and enables all kind of
+interaction with all resources available in etcd DB but its usage often requires
+to build verbose yaml files. Also kubectl tool only expose the standard K8s
```
`s/yaml/YAML:`

Ebaneck replied:
`s/Also/Also,`
`s/expose/exposes`

thomasdanan replied:
done

slaperche-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+to build verbose yaml files. Also kubectl tool only expose the standard K8s
+resources but not the custom ones, such as Volumes.
+
+Currently, MetalK8s provides other set of CLI or manual procedures, but those
```
`a set of tools` not CLI.

I don't think CLI means what you think. A CLI is defined as `A user interface in which the user interacts with the computer through lines of text instead of graphical images.`

What we have is a bunch of script that we can run, and they are not really interactive.
You could say that this bunch of scripts defines a CLI for MetalK8s, but a script by itself isn't a CLI.

thomasdanan replied:
done

slaperche-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+Because kubectl CLI is already in place and is well known by Kubernetes
+administrators, it will be used as a standard to follow for all other MetalK8s
+CLIs:
+- CLI follows kubctl style
```
`s/kubctl/kubectl/`

I may have missed some occurrences, but this typo is present several times (may be worth a search & replace).

thomasdanan replied:
done

slaperche-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,57 @@
+- CLI follows kubctl style
+- CLI is not interactive
+- CLI should not require password input
+- CLI implementation relies on secured APIs
```
Not sure about that, but I think `secure APIs` instead of `secured APIs`.

thomasdanan replied:
done

slaperche-scality reviewed (commented):
I think there is a misuse of the term CLI in the document.

thomasdanan committed:
requirements: fixup on high level principles
spelling issues
review completion principle
add help principle
CLI vs scripts and tools

thomasdanan committed:
Merge branch 'development/2.6' of https://github.com/scality/metalk8s into feature/2322-centralized-cli-specs

thomasdanan committed:
requirements: add req file to index
also fix formatting issues

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,62 @@
+
+- CLI follows kubectl style: kubectl <action> <object>
+- CLI provides an exhaustive help, per action, with relevant examples
+- CLI is not interactive (except maybe for very first install script)
```
> except maybe for very first install script

No for me it's really only for password and that's it. And this password should be ask at the very beginning of the command execution so that the user launch the command, enter the password if any and then not more interactivity

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,62 @@
+- CLI output is standardized and human readable by default
+- CLI output can be formatted in JSON or YAML
+- CLI can be executed from outside of the cluster (except maybe for very first
+  install script)
```
> except maybe for very first install script

No, not only. You also have for example the iso-manager, upgrade and downgrade script that need to run on the bootstrap node. For me you should remove this requirement or have a better "except" part but ... I don't know how you can define it for me it's something like "except for script that need to (re)start salt-master container" but it maybe not only this

thomasdanan replied:
@TeddyAndrieux the requirement is to be able to run everything from outside of the cluster, as we can do with kubectl. It is maybe not possible, you will say it in the design.

NicolasT replied:
I'm with @TeddyAndrieux here. Putting something in a requirement means "We really need this for some customer use-case". As such, either it's a must (requirement) and we need to figure out some design to accommodate the need, or it's not a requirement.

Now, in the case where 'local things' are required (indeed, e.g. ISO import, where one needs to refer to a *local* path to an ISO), *requiring* this to be available 'from outside' seems odd.

xaf-scality replied:
Well you can execute remotely some 'local things' with local path, and yes, it's a bit weird.
The point is why we don't want this ? Because of technical difficulties or conceptual things or security things ?
Moreover could the CLI upgrade itself ?

thomasdanan replied:
@NicolasT @xaf-scality @TeddyAndrieux let's discuss this one during today's meeting.

thomasdanan replied:
removed, only kubectl metalk8s plugin should be executable from outside of the cluster

thomasdanan committed:
start listing all possible commands

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,145 @@
++------------+------------+-----------------------------------------------+
+| bootstrap  | restore    | backup-file                                   |
++------------+------------+-----------------------------------------------+
+| bootstrap  | restore    | backup-file                                   |
```
2 time restore

thomasdanan replied:
done

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,145 @@
++------------+------------+-----------------------------------------------+
+| Resource   | action     | parameters                                    |
++============+============+===============================================+
+| cluster    | install    |                                               |
```
What `install cluster` mean ? It's bootstrap right ? Or it's something with a platform description file and this do bootstrap + expansions ?

Also I think we want a command to `install bootstrap`.

Since we use `<action> <resource>` style I think you should have `deploy` or `install` but not a mix because command will look like 
`kubectl install cluster`

thomasdanan replied:
you are right, it should be `install` `bootstrap`.
Not sure I understand your second comment

TeddyAndrieux replied:
For node you have `node` `deploy` and for bootstrap it's `bootstrap` `install` for me you should have only `deploy` or only `install`

thomasdanan replied:
okay understood

thomasdanan replied:
done

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,145 @@
++------------+------------------+-----------------------------------------+
+| node       | edit             | name, hostname (to change node hostname)|
++------------+------------------+-----------------------------------------+
+| node       | deploy(or apply?)| <list of nodes>, dry-run                |
```
Hmm it's not part of `kubectl`

thomasdanan replied:
isn't it the kind of thing we can achieve with the kubectl plugin? i.e. enrich kubectl with new functionalities? but from a usage point of view, the administrator will invoke kubectl?

TeddyAndrieux replied:
Hmm for this one is exacly one we do NOT want to put in a kubectl plugin because it's an interaction with salt directly so not part of kubectl at all

gdemonet replied:
(can't review the whole PR atm, but just my 2 cents on this question)

We **can** call SaltAPI from `kubectl`. We **could** also send commands in the `salt-master` container. Nothing prevents us from interacting with Salt from a `kubectl` plugin.

There are some commands which will require direct interaction on the host however, maybe because the operation will trigger a master restart (e.g. upgrade/downgrade), or because it does not make sense to do it from somewhere else (e.g. managing Metal or Solution ISOs on the Bootstrap). But deploying a Node is a standard operation you may want to do from outside the cluster (we do it from the UI already).

TeddyAndrieux replied:
Of course we can do it from a kubectl plugin but for me it's a bad idea. Kubectl plugin shouldn't interact with salt IMO.

> There are some commands which will require direct interaction on the host however, maybe because the operation will trigger a master restart (e.g. upgrade/downgrade), or because it does not make sense to do it from somewhere else (e.g. managing Metal or Solution ISOs on the Bootstrap). But deploying a Node is a standard operation you may want to do from outside the cluster (we do it from the UI already).

???
I do not see why it's related to kubectl plugin or not ?? You can execute your kubectl plugin from outside the cluster, you can also execute our own CLI from outside the cluster

gdemonet replied:
> you can also execute our own CLI from outside the cluster

No. This "extra CLI" we'll build for Bootstrap-local operations will not be a "redistributable package", just an executable in an ISO.

Otherwise, how do you allow this CLI to authenticate against Salt without a kubeconfig? Would you reimplement the login dance with K8s API to then be able to reach Salt API?
Also, what's the benefit of having two CLIs if they can both work anywhere?

TeddyAndrieux replied:
> No. This "extra CLI" we'll build for Bootstrap-local operations will not be a "redistributable package", just an executable in an ISO.

So for you all extra CLI are for stuff that cannot go outside the cluster otherwise it's a kubectl plugin ??
For me no.

To add a node you can perfectly have a CLI that work outside of the cluster or a kubectl plugin but IMO kubectl plugins shouldn't interact with salt

slaperche-scality replied:
@TeddyAndrieux That's the second time you say `IMO kubectl plugins shouldn't interact with salt`, do you have reasons/arguments for that?

I've no opinion on that part, but giving your argument for this could help convince people (or at least understand your point of view).

NicolasT replied:
Why not, @TeddyAndrieux ? There's SaltAPI they can work with, using the `KubeConfig` token.

Now, to be clear: `kubectl` plugins *only allow to add top-level commands*. So you *can't* write a plugin which all of a sudden would give you a `kubectl node deploy` command or something along those lines. You could have `kubectl metalk8s deploy node`.

NicolasT replied:
OK, in the above: you *could* have `kubectl node deploy` because `kubectl node` doesn't exist., but that's mere coincidence and I think it'd be very bad style (and potential future-incompatibility) to hijack random nouns.

TeddyAndrieux replied:
For me kubectl plugins should only act on apiserver and not on something we deploy on our side, for me it's something that can work on every kubernetes deployment ...

I agree and I know everything is doable from kubectl plugin but for me it's not a good idea. Anyway as you want

thomasdanan replied:
okay so If I understand correctly, with kubectl, we can add a new action but we can't override an existing one (like kubectl create node, but with additional arguments)?

gdemonet replied:
> for me it's something that can work on every kubernetes deployment 

Not really. Core `kubectl` functionalities, yes. But **plugins** are "additional functionality". For instance, we could have a `kubectl-prom` plugin which helps manage whatever comes with the Prometheus Operator.

> with kubectl, we can add a new action but we can't override an existing one (like kubectl add node, but with additional arguments)?

The way plugins work is quite simple. Put a binary called `kubectl-<plugin-name>` in the right place, and whenever you will run `kubectl <plugin-name> ...`, the call will be forwarded to this binary by `kubectl`, nothing more.
So indeed, there's no way to override existing `kubectl` functionality with plugins, only to extend it. In our case, as Nicolas said, we'll likely have all our commands under `kubectl metalk8s ...`.

thomasdanan replied:
Okay, thanks for the explanation ... I think I understand now

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,145 @@
+tools: kubectl
+
++------------+------------------+-----------------------------------------+
+| Resource   | action           | parameters                              |
```
If we follow kubectl style I'm not sure it's good to describe it like this. 

I think you should list for each `action` what `resource` you can use. 

BUT from my understanding we want several different `tool` (at least 2) so that we have one that can be run from anywhere with the right kubeconfig and another that need to be on the bootstrap node  cc @NicolasT

thomasdanan replied:
We want all commands to behave the same, i.e. <tool> <action> <resource>. But in terms of requirements, I prefer to express the list of tasks we want to expose through this CLI and organize the tasks per domain.

I am not sure to understand why we should enforce the fact that some tools must run on the bootstrap. It looks like more a design constraint than a requirement no? Maybe at the end, we will say, some actions cannot be run from outside of the cluster and that's okay ...

TeddyAndrieux replied:
True it's not part of the requirement directly

thomasdanan committed:
change install command target resource (bootstrap)
interactive req only if passwd required

thomasdanan committed:
Merge branch 'development/2.6' of https://github.com/scality/metalk8s into feature/2322-centralized-cli-specs

thomasdanan committed:
centralized cli reqs: ready for review version
- list all actions we can find in the doc
- remove install action (will always use deploy)
- describe the vision with 2 CLI: kubectl and mk8sctl

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+All operations that need to be performed by the Platform Administrator could be
+categorized as follow:
+
+- Cluster Resources Administration (Nodes, Volumes, Deployments, ...)
```
What are *Deployments* in this context?

xaf-scality replied:
Deployments means Expansions ?

thomasdanan replied:
I was referring to Deployment Object on K8s API

NicolasT replied:
That's kinda confusing because MetalK8s doesn't really 'care' that much about (K8s) *Deployments*. I'd suggest another term ('cluster services'?)

gdemonet replied:
Maybe 'addons'?

thomasdanan replied:
Actually, I am talking about any K8s resource we can create, edit, delete through k8s apiserver. I suggest k8s resource

gdemonet replied:
But this should be left to `kubectl` as we said... Do you still want to mention this in the requirements?

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+- Cluster Service Administration (Configure Dex, Prometheus, Alert Manager,
+  ...)
+
+K8s provides the kubectl CLI, which is distributed and enables
```
*distributed*?

thomasdanan replied:
Indeed, not the right term, wanted to say that we can execute it anywhere since we have Network access to the cluster ...

thomasdanan replied:
done

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+- CLI is not interactive (except if password input is needed)
+- CLI should not require password input
+- CLI provides a dryrun mode for intrusive operations
+- CLI provides a verbose (or debug) mode
```
Note, `kubectl` verbosity is not a toggle but a value.

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+- CLI support <action> completion for easy discovery
+- CLI output is standardized and human readable by default
+- CLI output can be formatted in JSON or YAML
+- CLI can be executed from outside of the cluster (except maybe for very first
```
This seems to imply there's a single CLI (tool) somewhere. I don't think that's (necessarily) the intent?

xaf-scality replied:
The intent is to have less # of CLI, so the more we can aggregate in one CLI the best it is.

thomasdanan replied:
@NicolasT my suggestion is to have, in addition to kubectl, one other CLI: mk8sctl for all platform operations. Note that with solutions, we may have additional CLI for Zenko, HD and SOFS. I think it makes sense to limit the number of CLI we expose.

gdemonet replied:
Solutions could provide their own `kubectl` plugin ;)

thomasdanan replied:
smart !!!

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+
+All functionalities are exposed through 2 distincts CLI:
+- kubectl (enriched using kubectl approach)
+- mk8sctl: a new CLI, exposing specific MetalK8s functionalities
```
`metalk8sctl` or similar, please... We aim for tab-completion of actions, so why not use a shell's tab-completion.

thomasdanan replied:
done

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+
+When it is possible, it would make sense to leverage kubectl plugin
+
+All functionalities are exposed through 2 distincts CLI:
```
Here you say *all* functionality should be exposed using *both* a `kubectl` extension *as well as* some other tool. That seems odd, unintuitive and confusing.

xaf-scality replied:
I agree why having two CLI that will do the same ?

thomasdanan replied:
@NicolasT can we do everything with kubectl and kubectl plugin?

thomasdanan replied:
@xaf-scality the 2 commands are not doing the same things but again if we can do everything with kubectl that's even better ...

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+- kubectl (enriched using kubectl approach)
+- mk8sctl: a new CLI, exposing specific MetalK8s functionalities
+
+In order to operate the cluster with mk8sctl from outside of the cluster, a
```
- For any `kubectl` extra functionality, some binary/binaries need to be installed anyway as well.
- I'd say: `metalk8scli` for on-node actions, `kubectl` plugin for anything that can be done remotely.

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+In order to operate the cluster with mk8sctl from outside of the cluster, a
+specific pkg (for each OS) or a procedure explaining how to deploy it is
+available.
+The mk8scli is deployed and available by default on bootstrap nodes
```
As well as the `kubectl` plugin, then.

thomasdanan replied:
done

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
++------------+------------------+-----------------------------------------+
+| Resource   | action           | parameters                              |
++============+==================+=========================================+
+| node       | add              | name, ssh-user, hostname or ip,         |
```
That's not how it works. It'll be something like `kubectl metalk8s create node ...`

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
++------------+------------------+-----------------------------------------+
+| Resource   | action           | parameters                              |
++============+==================+=========================================+
+| node       | add              | name, ssh-user, hostname or ip,         |
```
The interactive action of `kubectl` is `create`, not `add`.

thomasdanan replied:
done

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+|            |                  | ssh port, ssh-key-path, sudo-required,  |
+|            |                  | roles, manifest-template-path           |
++------------+------------------+-----------------------------------------+
+| the manifest-template-path is the path to the manifest describing the   |
```
That's not in line with other `kubectl` commands. Why would one want to provide such manifest?

thomasdanan replied:
@NicolasT, in order to simplify the number of inputs a user should provide in order to perform a specific operation.

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+| resource. Everything that is in the command line, override the manifest |
+| content. This apply to Node and Volumes.                                |
++------------+------------------+-----------------------------------------+
+| node       | edit             | name, hostname (to change node hostname)|
```
No. `kubectl edit` opens an editor to interactively edit an object. One can already `kubectl edit node ...` and do whatever.

thomasdanan replied:
@NicolasT my problem is when we have to edit such a yaml file, it is super scary, it is error prone, and unless you are familiar with kubernetes and yaml, you will be very likely unable to perform your operation.

gdemonet replied:
We could indeed have a "simplified" node object with just the few fields we know are editable. But that's quite a lot of effort, whereas we could expose the `set` verb that `kubectl` has for some resouces.

Something like `kubectl metalk8s set ssh-host node/master-1 10.200.1.2`.

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+| resource. Everything that is in the command line, override the manifest |
+| content. This apply to Node and Volumes.                                |
++------------+------------------+-----------------------------------------+
+| node       | edit             | name, hostname (to change node hostname)|
```
What would 'changing the hostname' do? You can't change the `name` of an object in K8s...

thomasdanan replied:
@NicolasT actually it will do all of this: https://metal-k8s.readthedocs.io/en/latest/operation/changing_node_hostname.html

gdemonet replied:
Oh, that's not "changing a hostname". This procedure is basically replacing a Node entirely (it's not even draining it beforehand). Not sure we want to expose this as a command, especially not one that makes you believe it's "just changing a value in some YAML".

thomasdanan replied:
will remove this one

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+| node       | delete, drain,   | <list of nodes>                         |
+|            | taint, label     |                                         |
++------------+------------------+-----------------------------------------+
+| volume     | add              | name, nodeName, storageClassName,       |
```
This (`devicePath` and such) is too tied to the current type of volume we may support (and only one of the two we do support). This needs to be revised into a proper CLI UX.

xaf-scality replied:
this one for CLI will be a very long command line with all the "optional values" for each type of volume. It's almost like writing the whole yaml file describing the volume in this case

thomasdanan replied:
@xaf-scality @NicolasT What I was suggesting is to have manifest templates and being able to override some values in command line ... to be discussed during today's meeting

thomasdanan replied:
will remove this concept of manifest template

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+| bootstrap  | restore    | backup-file                                   |
++------------+------------+-----------------------------------------------+
+| cluster    | edit-conf  | conf-param=conf-value                         |
+|            |            | ex: ip-in-ip=true                             |
```
I'm pretty sure you'll run into trouble with this.

`BootstrapConfiguration` is a structured object, which here you're flattening into... what? Which subactions would you support?...

What you'll want instead is a config check (is the YAML valid?), and a (re)deploy-with-new-config.

thomasdanan replied:
will remove this one

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
++------------+------------+-----------------------------------------------+
+| Resource   | action     | parameters                                    |
++============+============+===============================================+
+| solution   | import     |                                               |
```
Without a path to some ISO?

thomasdanan replied:
done

NicolasT commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
++------------+------------+-----------------------------------------------+
+| Resource   | action     | parameters                                    |
++============+============+===============================================+
+| grafana    | edit-conf  | conf-param=conf-value                         |
```
This (and the ones below) won't work. If there's structured editing going on, then present the user with an editor.

thomasdanan replied:
done

xaf-scality commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,172 @@
+CLIs:
+
+- CLI follows kubectl style: kubectl <action> <resource>
+- CLI provides an exhaustive help, per action, with relevant examples
```
auto completion for actions (and not values) like ringsh ?

thomasdanan committed:
requirements: fixup
- remove distributed term which was not appropriate
- fix spelling issues
- fix commands args

thomasdanan committed:
requirements: fixup
- change mk8sctl in metalk8scli
- change add into create action

thomasdanan committed:
centralized cli requirements: major re edit
- freeze the approach with 2 CLI and kubectl plugin
- remove commands already supported by native kubectl
- remove the kubectl style req, since metalk8sctl will behave 
differently

thomasdanan commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux







*The following options are set:* **approve**

TeddyAndrieux commented on docs/developer/architecture/centralized-cli.rst:
```diff
@@ -0,0 +1,151 @@
+- kubectl: enriched with metalk8s plugin, to interact with both k8s apiserver
+  and salt API, and that can be executed from outside of the cluster.
+- metalk8sctl: a new CLI, exposing specific MetalK8s functionalities, that are
+  not interacting with k8s apiserver, and that must be executed on cluster node
```
> that are not interacting with k8s apiserver

It's not true for me, some of them interact with apiserver just they must be run on the machine because they interact also with local machine

thomasdanan replied:
ok, will do that in one other PR if you don't mind

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye thomasdanan.
